### PR TITLE
Insert multiple rows using a single SQL statement

### PR DESCRIPTION
**Pull request by jkugiya**
It solves https://github.com/slick/slick/issues/1272.

Another possible implementation is [here](https://github.com/jkugiya/slick/pull/1), but i think this is better approch. (Building SQL are on `JdbcStatementBuilderComponent`)

- ~~I haven't fully investigated how to handle the return value when there are multiple values. It could be `MultiInsertResult` as well as `++=`.~~
  - It returns `MultiInsertResult`, but it may returns Nil if the driver doesn't support returning each ids.
- I think this implementation can be applied to the implementation of upsert with multiple values. On the other hand, I think we need to consider how we should name it in that case. (Uniform naming between insert and upsert is desirable, I have also implemented batch upsert #2397).
  1. The API signature will be a little more difficult, following is also possible.
    ```scala
    def ++= (values: Iterable[T])(implicit im: InsertMode = InsertMode.BatchInsert): ProfileAction[MultiInsertResult, NoStream, Effect.Write] = 
      im match {
        case InsertMode.BatchInsert => batchInsertAll(values)
        case InsertMode.MultipleValues =>insertAllByMultipleValues(values)
      }
    ```
  2. Adding new methods, simply
    ```scala
  // use alias
  def ++=(values: Iterable[T]): ProfileAction[MultiInsertResult, NoStream, Effect.Write] = insertAllBatch(values)
  // batch insert
  def insertAllBatch(values: Iterable[T]): ProfileAction[MultiInsertResult, NoStream, Effect.Write]
  // multiple rows 
  def insertAll(values: Iterable[T]): ProfileAction[MultiInsertResult, NoStream, Effect.Write]
  def insertOrUpdateAllBatch(values: Iterable[T]): ProfileAction[MultiInsertResult, NoStream, Effect.Write]
  def insertOrUpdateAll(values: Iterable[T]): ProfileAction[MultiInsertResult, NoStream, Effect.Write] 
    ```

### CODE REVIEW

**Comment by nafg** [comment]
I added a workflow to report mima results on the PR, closing & reopening to trigger

**Comment by nafg** [comment]
Breaking changes are okay, but when applicable should be documented in [upgrade.md](../blob/main/doc/paradox/upgrade.md)

**Comment by nafg** [comment]
Hi what's the status of this?

**Comment by jkugiya** [comment]
@nafg 

1. A decision needs to be made whether or not to override the existing `++=` implementation.  
  For this I prefer to make it an alias method since the behavior will change.
2. If it is to be an alias method, I would like to add a similar method to ``insertOrUpdate``(#2397 ), so I need a regular naming convention. I have written a proposal in the description and would appreciate your input.

**Comment by nafg** [comment]
When would one prefer, or be relying on, the old behavior? Isn't this simply an optimization?

**Comment by nafg** [comment]
I would not use an implicit mode. If there is a justification to keep both modes (as separate methods, whether or not we deprecate the old mode), I think a good naming convention would be to suffix with Batch (e.g. insertAllBatch).

Another option would be a Boolean `batch` parameter with a default value.

**Comment by nafg** [comment]
Not sure I fully understood the description.

Also can you list pros and cons of this versus the other approach, or are you settled on this one already?

File: doc/paradox/upgrade.md
Comment (nafg) [inline doc/paradox/upgrade.md:37]: Doesn't Slick already have a few such things? How does this differ?

**Comment by nafg** [comment]
Threading around Int offsets seems brittle but if that's the only obvious way that isn't too much work then it is what it is

File: doc/paradox/upgrade.md
Comment (jkugiya) [reply doc/paradox/upgrade.md:37]: I have modified it. This is change log.

Comment (jkugiya) [reply doc/paradox/upgrade.md:37]: >  If there is a justification to keep both modes (as separate methods, whether or not we deprecate the old mode), I think a good naming convention would be to suffix with Batch (e.g. insertAllBatch).

I think there is a reason to have both methods. The existing method has performance issues, but requires a smaller communication buffer. On the other hand, if there are inserting a very large number of values, the new method may be prone to communication errors because of its large sql statements.

**Comment by jkugiya** [comment]
>  If there is a justification to keep both modes (as separate methods, whether or not we deprecate the old mode), I think a good naming convention would be to suffix with Batch (e.g. insertAllBatch).

I think there is a reason to have both methods. The existing method has performance issues, but requires a smaller communication buffer. On the other hand, if there are inserting a very large number of values, the new method may be prone to communication errors because of its large sql statements.

**Comment by jkugiya** [comment]
> I would not use an implicit mode. 

While implicit is useful for changes to existing code, since the new method can be applied simply by importing mode, if this is not desirable as an API design, it seems like it would be better to simply provide an alias method. For the boolean flag, all the codes for the infix notation must be changed.

File: doc/paradox/upgrade.md
Comment (nafg) [reply doc/paradox/upgrade.md:37]: Yes but maybe make it clear why it was added given that there are other similar things

**Comment by nafg** [comment]
Please rebase

**Comment by lightbend-cla-validator** [comment]
Hi @jkugiya,

Thank you for your contribution! We really value the time you've taken to put this together.

We see that you have signed the Lightbend Contributors License Agreement before, however, the CLA has changed since you last signed it.
Please review the new CLA and sign it before we proceed with reviewing this pull request:

[https://www.lightbend.com/contribute/cla](https://www.lightbend.com/contribute/cla)

**Comment by lightbend-cla-validator** [comment]
Hi @jkugiya,

Thank you for your contribution! We really value the time you've taken to put this together.

We see that you have signed the Lightbend Contributors License Agreement before, however, the CLA has changed since you last signed it.
Please review the new CLA and sign it before we proceed with reviewing this pull request:

[https://www.lightbend.com/contribute/cla](https://www.lightbend.com/contribute/cla)

**Comment by nafg** [comment]
@jkugiya I want to tackle this now and merge it. Could we have a voice chat perhaps?

**Comment by jkugiya** [comment]
> When would one prefer, or be relying on, the old behavior? Isn't this simply an optimization?

It is not a simple optimization. Batch inserts and bulk inserts each have their own drawbacks and advantages.

- batch insert
  - Pros
     - The small size of the data sent makes errors due to buffer size less likely to occur
   - Cons
     - It is inefficient because the number of times SQL is sent is very high.
- Bulk insert(multi row insert)
  - Pros
    - Efficient
  - Cons
    - Depending on the size of the SQL, an error may occur regarding the buffer. This may affect products that assume the current implementation.

File: doc/paradox/upgrade.md
Comment (jkugiya) [reply doc/paradox/upgrade.md:70]: > In order to support issuing insert statements with multiple values in Slick 3.4
(`++=` is a batch call of multiple single insert statement), we have made this interface available for offsets.
If you have extended a profile and used the `set` method of this interface,
you should specify 0 for this offset unless the number of variable placeholders in the SQL dynamically changes.

The reason is this, but is it not clear?

File: slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala
Comment (jkugiya) [inline slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:373]: I thought that there are not always only two options for optimization, so I decided to specify an enum instead of a `Boolean` type.

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/InsertTest.scala
Comment (jkugiya) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/InsertTest.scala:94]: https://github.com/slick/slick/blob/main/slick/src/main/scala/slick/jdbc/DB2Profile.scala#L38-L42

There seems to be a bug in the DB2 driver that when using `returning` to return a column value, the column name must be uppercase.

File: slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala
Comment (nafg) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:373]: Great, that's better than a boolean. But unless this has applicability for more statements than just insert, I would make its name more specific. So maybe `InsertAllMode` or something.

Also, the naming is inconsistent -- why not `BatchStatement` or `Single`? I would omit the `Statement` suffix since it would be obvious because you're dotting into it (`StatementOption.Single`).

Also, since batch might not be supported perhaps the second option is misleading?

Actually since "MultiInsert" is a term that's used already, how about `MultiInsertStrategy.EachRow` and `.AllRows`?

I can make this change on Sunday if you like it and don't get to it first.

Another thought, since it's just a data type we could have another option:

```scala
    case class ChunkedRows(size: Int) extends MultiInsertStrategy
```

which would mean to do `size` rows per statement, so it would use multi-row VALUES clauses but if the number of rows to insert exceeds `size` it will still generate multiple statements. You mentioned the tradeoff, that doing them all in a single statement could cause issues sometimes. Having this option could allow optimizing the tradeoff.

We could implement this by giving `MultiInsertStrategy` a `def size: Int` and `EachRow` would override it to `1` and `AllRows` (or whatever we name them) would override it to `Int.MaxValue`. Then the implementation wouldn't have to pattern match. It could just read `size`.

That said, at that point they are isomorphic to their `size` value. Which means potentially instead of introducing a new data type and using it as the parameter, we could simply have a `rowsPerStatement: Int = Int.MaxValue` parameter.

File: slick/src/main/scala/slick/jdbc/OracleProfile.scala
Comment (nafg) [inline slick/src/main/scala/slick/jdbc/OracleProfile.scala:547]: I don't like runtime errors.

Is there any way to design it that only the right databases would have this API?

If you don't get to this by Sunday I'll try to work on that then.

File: slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala
Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:373]: > So maybe InsertAllMode or something.

I was wondering if I need to take care of insertOrUpdate, but I am not particular about it.

Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:373]: > Also, the naming is inconsistent -- why not BatchStatement or Single? I would omit the Statement suffix since it would be obvious because you're dotting into it (StatementOption.Single).

Batch/Single looks good. (To be precise, Batch is the use of JDBC functions, and Single is the SQL syntax.)

Comment (nafg) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:373]: > > So maybe InsertAllMode or something.
> 
> I was wondering if I need to take care of insertOrUpdate, but I am not particular about it.

Maybe a better naming scheme is `RowsPerStatement.{One, All}` (meaning the data type would be called `RowsPerStatement` and the choices would be `One` and `All`)

Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:373]: > Actually since "MultiInsert" is a term that's used already, how about MultiInsertStrategy.EachRow and .AllRows?

My hope is that the API users can guess whether they are using JDBC functionality (= batch) or SQL statement. It seems a little too abstract in that regard.

Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:373]: > ChunkedRows

It would be nice to use such an interface as a first improvement.
However, after looking at it a bit, there are several options for future support of Oracle(https://database.guide/4-ways-to-insert-multiple-rows-in-oracle/), and this may be the case with other profiles. Therefore, I feel it is better to use enum as an interface than a simple Int type, even if it is only `ChunkedRows` at first.

File: slick/src/main/scala/slick/jdbc/OracleProfile.scala
Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/OracleProfile.scala:547]: I thought that using macros might help (I haven't tried it at all), but since Scala3 support is coming up, I didn't adopt that idea this time.

Comment (nafg) [reply slick/src/main/scala/slick/jdbc/OracleProfile.scala:547]: No the solution is to have more precise types. Sorry I didn't get to it yet

**Comment by nafg** [comment]
@jkugiya please review my changes, if you're happy then I'm happy.

I squashed your changes into two commits, and my changes are a few small commits on top of yours. If you approve I'll squash my changes similarly.

File: slick/src/main/scala/slick/jdbc/OracleProfile.scala
Comment (nafg) [reply slick/src/main/scala/slick/jdbc/OracleProfile.scala]: I've implemented this as https://github.com/slick/slick/commit/7e60b4369e1c6aa2be3ab80c4e4331ded37465de -- thoughts?

I would do it as a separate PR.

File: slick/src/main/scala/slick/jdbc/RowsPerStatement.scala
Comment (jkugiya) [inline slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:3]: I think what we should show in options is not how many lines to insert, but how they are implemented.

Currently implementation for Oracle is not supported in this PR, but it may need to provide more than one of several possible implementations.

https://database.guide/4-ways-to-insert-multiple-rows-in-oracle/

File: slick/src/main/scala/slick/jdbc/RowsPerStatement.scala
Comment (jkugiya) [inline slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:10]: I understand that this is generally a good way to provide type safety in Scala.
However, each Profile may have different options that should be supported.
I believe Slick has an architecture that basically provides APIs in JdbcProfile and different ones for different database products are provided in each Profile.
This implementation forces each Profile to have only two types of implementations.

**Comment by jkugiya** [comment]
@nafg 

I left two comments, neither of which is a very strong opinion. I think there could be a decision to not dare to be so flexible.

File: slick/src/main/scala/slick/jdbc/RowsPerStatement.scala
Comment (nafg) [reply slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:3]: > I think what we should show in options is not how many lines to insert, but how they are implemented.

I'm not clear what you're suggesting by this (or what it has to do with the rest of the comment)

File: slick/src/main/scala/slick/jdbc/RowsPerStatement.scala
Comment (nafg) [reply slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:10]: I achieved this in https://github.com/slick/slick/commit/7e60b4369e1c6aa2be3ab80c4e4331ded37465de (which I could add to the branch, or PR separately after this is merged) by using a type member.

https://github.com/slick/slick/commit/7e60b4369e1c6aa2be3ab80c4e4331ded37465de#diff-fef81a55b3619e8e8d49ed29a91ec391c8a7d87e76782212f885c30958786822

It's true that even there, I did it to accommodate Oracle specifically, not in a fully open way, but it could be tweaked a little to achieve that.

tl;dr type members + abstract methods allow refining a subset of a sealed trait hierarchy.

That said I don't recall adding `sealed`, so I'm unclear if your critique is of `sealed`, of my having moved the hierarchy to top-level, or something else.

I moved it to top-level to make imports easier and more consistent with other things in Slick.

**Comment by nafg** [comment]
@jkugiya I replied on the comments. Let me know what changes exactly would make you happy with me merging it.

We can always tweak it more in another PR. I don't want to merge your PR unless we're both happy with it. OTOH if there is a small tweak that could be a separate PR even if it will have discussion and we merge this, it might be better for working on other things with less concern of merge conflicts...

Either way I won't merge until I get a clear go ahead from you.

File: slick/src/main/scala/slick/jdbc/RowsPerStatement.scala
Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:10]: Oh, I thought I had removed sealed, but it remained.

File: slick/src/main/scala/slick/jdbc/RowsPerStatement.scala
Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:3]: My feeling is that the number of rows is just not a very important concept.
The difference of these options is whether the implementation uses JDBC's batch API or the syntax of each database product. In the future, when Oracle is supported, there may be a difference between using INSERT ALL and Select Query.

Thus, I had the following names in mind as candidates

- Options
  - InsertOptions
  - StatementOptions
- Behaviors
- Mode
  - StatementMode

Comment (nafg) [reply slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:3]: > My feeling is that the number of rows is just not a very important concept.

I mean it is, it's the whole reason for this PR.

> The difference of these options is whether the implementation uses JDBC's batch API

That's not correct. Even with one-row-per-statement whether or not batching is used is a separate decision that's not controlled by this flag. It may not batch. The only thing that's certain in that case is that each row will involve a separate SQL statement.

> or the syntax of each database product. In the future, when Oracle is supported, there may be a difference between using INSERT ALL and Select Query.

I **think** you're alluding to the possibility of using this ADT to control which of the different ways of doing a single insert with all the rows in Oracle. I don't see the point of that. If we add support for single insert for Oracle we can just pick one method. Unless there are significant trade-offs between the different ways of achieving it, I don't think the user cares. But they do care that it's one insert statement with all the rows.

Plus, as discussed we may want to add an option for N rows per statement. So I really think it's about rows per statement, not anything else.

> 
> Thus, I had the following names in mind as candidates
> 
> * Options
>   
>   * InsertOptions
>   * StatementOptions
> * Behaviors
> * Mode
>   
>   * StatementMode

I feel like those are too vague. It's not obvious what they're for, which means users have to work harder to learn it, and library contributors have to work harder to not have name clashes and to know how to evolve things later.

Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:3]: I understand your opinion, but my opinion remains the same.
However, it is only a matter of name, so either way is fine.

Comment (jkugiya) [reply slick/src/main/scala/slick/jdbc/RowsPerStatement.scala:3]: > that's not controlled by this flag. It **may not** batch. 

Please understand that this is a story that could change the behavior of the existing ++=(= batch for each row).  All existing products using Slick are running in batch mode in all Profiles.

**Review by nafg — APPROVED** [review]

**Pull request merged**
